substrate/state-machine/src/lib.rs

```diff
@@ -257,46 +261,65 @@ pub fn execute_using_consensus_failure_handler<
 	manager: ExecutionManager<Handler>,
 ) -> Result<(Vec<u8>, B::Transaction), Box<Error>> {
 	let strategy: ExecutionStrategy = (&manager).into();
+
+	// make a copy.
+	let code = ext::Ext::new(overlay, backend).storage(b":code")
+		.ok_or(Box::new(ExecutionError::CodeEntryDoesNotExist) as Box<Error>)?
+		.to_vec();
+
 	let result = {
-		let maybe_orig_prospective =
-			if let &ExecutionManager::Both(_) = &manager { Some(overlay.prospective.clone()) } else { None };
-
-		let ((result, was_native), code, delta) = {
-			let mut externalities = ext::Ext::new(overlay, backend);
-			// make a copy.
-			let code = externalities.storage(b":code")
-				.ok_or(Box::new(ExecutionError::CodeEntryDoesNotExist) as Box<Error>)?
-				.to_vec();
-
-			(
-				exec.call(
-					&mut externalities,
-					&code,
-					method,
-					call_data,
-					// attempt to run native first, if we're not directed to run wasm only
-					strategy != ExecutionStrategy::AlwaysWasm,
-				),
-				code,
-				externalities.transaction()
-			)
+		let mut orig_prospective = overlay.prospective.clone();
+
+		let (result, was_native, delta) = loop {
+			let ((result, was_native), delta) = {
+				let mut externalities = ext::Ext::new(overlay, backend);
+				(
+					exec.call(
+						&mut externalities,
+						&code,
+						method,
+						call_data,
+						// attempt to run native first, if we're not directed to run wasm only
+						strategy != ExecutionStrategy::AlwaysWasm,
+					),
+					externalities.transaction()
+				)
+			};
+
+			if result.as_ref().err().map_or(false, |e| e.needs_retry()) {
+				overlay.prospective = orig_prospective.clone();
```

Comment:
this clone seems unnecessary and quite heavy

Reply:
most of the time `orig_prospective` will be empty. it's necessary since `overlay.prospective` gets altered as part of calling into runtime.